src/libertem/utils/threading.py

```diff
@@ -138,15 +200,30 @@ def set_num_threads(n):
     import scipy  # noqa: F401
     import numpy as np  # noqa: F401
     global __threadpool_wrapper
-    if __threadpool_wrapper is None:
-        __threadpool_wrapper = ThreadpoolWrapper()
+    global __threadpool_counter
+    global __threadpool_limiter
+
+    with lock:
+        if __threadpool_wrapper is None:
+            __threadpool_wrapper = ThreadpoolWrapper()
     # We use __threadpool_wrapper last so that it can cover
     # libraries that the other ones load
     with set_fftw_threads(n):
         with set_torch_threads(n):
             with set_numba_threads(n):
-                with __threadpool_wrapper(n):
+                try:
+                    with lock:
+                        if __threadpool_counter == 0:
+                            __threadpool_limiter = __threadpool_wrapper(n)
+                            __threadpool_limiter.__enter__()
+                        __threadpool_counter += 1
                     yield
+                finally:
+                    with lock:
+                        __threadpool_counter -= 1
+                        if __threadpool_counter == 0 and __threadpool_limiter is not None:
+                            __threadpool_limiter.__exit__(None, None, None)
```

Comment:
@sk1p Here I am not sure what to do. `None, None, None` would be only the "happy case". What should we do if the `finally:` comes after an exception?

Reply:
Hmm. I guess this will work with the current implementation of `_ThreadpoolLimiter.__exit__`, which just ignores any exceptions. I also can't currently make up a reason why it would behave differently in the case of exceptions, but maybe I'm not thinking creatively enough :) Instead of using the context manager API of `_ThreadpoolLimiter`, can we explicitly call `restore_original_limits` instead? It seems to be a quasi-stable interface, with an original name `_ThreadpoolLimiter.unregister` being kept as a backwards-compatible alias.

Thoughts?

Reply:
Ok, sounds good! I'm testing it locally now.

Reply:
As a justification for using the almost-internal API: at least we will see the breakage, which is not the case with the nesting approach...